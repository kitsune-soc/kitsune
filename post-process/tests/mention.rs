use pest::Parser;
use post_process::{PostParser, Rule};
use pretty_assertions::assert_eq;

#[test]
fn invalid_mention() {
    let text = "@test@hello.world@tes";
    assert!(PostParser::parse(Rule::mention, text).is_err());

    let text = "@ test@hello.world";
    assert!(PostParser::parse(Rule::mention, text).is_err());

    let text = "@test@hello.world@tes hello";
    let mut token_iter = PostParser::parse(Rule::post, text).unwrap();
    let text1 = token_iter.next().unwrap();
    assert_eq!(text1.as_rule(), Rule::text);
    assert_eq!(text1.as_str(), "@test@hello.world@tes hello");
    assert!(token_iter.next().is_none());

    let text = "@\\test@hello.world";
    assert!(PostParser::parse(Rule::mention, text).is_err());

    let text = "@test@hel\\lo.world";
    assert!(PostParser::parse(Rule::mention, text).is_err());
}

#[test]
fn subdomain_mention() {
    let text = "@test@social.hello.world";
    let mut token_iter = PostParser::parse(Rule::mention, text).unwrap();
    let mention = token_iter.next().unwrap();
    assert_eq!(mention.as_str(), "@test@social.hello.world");
}

#[test]
fn parse_mention() {
    let text = "hello @桐生@friday.night @真島! ";
    let mut token_iter = PostParser::parse(Rule::post, text).expect("Failed to parse post");

    let text1 = token_iter.next().unwrap();
    assert_eq!(text1.as_rule(), Rule::text);
    assert_eq!(text1.as_str(), "hello");

    let mention1 = token_iter.next().unwrap();
    assert_eq!(mention1.as_rule(), Rule::mention);
    assert_eq!(mention1.as_str(), " @桐生@friday.night");

    let mut mention1 = mention1.into_inner();
    let mention1_prefix = mention1.next().unwrap();
    assert_eq!(mention1_prefix.as_rule(), Rule::component_prefix);
    assert_eq!(mention1_prefix.as_str(), " ");
    let mention1_username = mention1.next().unwrap();
    assert_eq!(mention1_username.as_rule(), Rule::mention_username);
    assert_eq!(mention1_username.as_str(), "桐生");
    let mention1_domain = mention1.next().unwrap();
    assert_eq!(mention1_domain.as_rule(), Rule::mention_domain);
    assert_eq!(mention1_domain.as_str(), "friday.night");

    let mention2 = token_iter.next().unwrap();
    assert_eq!(mention2.as_rule(), Rule::mention);
    assert_eq!(mention2.as_str(), " @真島");
    let mut mention2 = mention2.into_inner();
    let mention2_prefix = mention2.next().unwrap();
    assert_eq!(mention2_prefix.as_rule(), Rule::component_prefix);
    assert_eq!(mention2_prefix.as_str(), " ");
    let mention2_username = mention2.next().unwrap();
    assert_eq!(mention2_username.as_rule(), Rule::mention_username);
    assert_eq!(mention2_username.as_str(), "真島");

    let text2 = token_iter.next().unwrap();
    assert_eq!(text2.as_rule(), Rule::text);
    assert_eq!(text2.as_str(), "! ");
}
