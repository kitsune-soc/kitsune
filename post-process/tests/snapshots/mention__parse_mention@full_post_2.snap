---
source: post-process/tests/mention.rs
expression: "PostParser::parse(Rule::post, &post).unwrap()"
input_file: post-process/tests/input/mention/full_post_2
---
[
    Pair {
        rule: mention,
        span: Span {
            str: ">@kiryu",
            start: 0,
            end: 7,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: ">",
                    start: 0,
                    end: 1,
                },
                inner: [],
            },
            Pair {
                rule: mention_username,
                span: Span {
                    str: "kiryu",
                    start: 2,
                    end: 7,
                },
                inner: [],
            },
        ],
    },
    Pair {
        rule: text,
        span: Span {
            str: " ",
            start: 7,
            end: 8,
        },
        inner: [],
    },
    Pair {
        rule: mention,
        span: Span {
            str: "<@akiyama@shun.me",
            start: 8,
            end: 25,
        },
        inner: [
            Pair {
                rule: component_prefix,
                span: Span {
                    str: "<",
                    start: 8,
                    end: 9,
                },
                inner: [],
            },
            Pair {
                rule: mention_username,
                span: Span {
                    str: "akiyama",
                    start: 10,
                    end: 17,
                },
                inner: [],
            },
            Pair {
                rule: mention_domain,
                span: Span {
                    str: "shun.me",
                    start: 18,
                    end: 25,
                },
                inner: [],
            },
        ],
    },
    Pair {
        rule: text,
        span: Span {
            str: " allo",
            start: 25,
            end: 30,
        },
        inner: [],
    },
]
